Make `std::mem::transmute_copy` accept `?Sized` inputs

nvzqz opened the pull request:
This enables conversions from dynamically-sized types like `[u8]`.

nvzqz committed:
Make `std::mem::transmute_copy` accept `?Sized` inputs

This enables conversions from dynamically-sized types like `[u8]`.

nvzqz commented:
@rustbot label +T-libs-api -T-libs

nvzqz committed:
Add `std::mem::transmute_copy` slice test

workingjubilee commented on library/core/src/mem/mod.rs:
```diff
@@ -1052,14 +1052,14 @@ pub const fn copy<T: Copy>(x: &T) -> T {
-pub const unsafe fn transmute_copy<Src, Dst>(src: &Src) -> Dst {
+pub const unsafe fn transmute_copy<Src: ?Sized, Dst>(src: &Src) -> Dst {
     assert!(
-        size_of::<Src>() >= size_of::<Dst>(),
+        size_of_val(src) >= size_of::<Dst>(),
         "cannot transmute_copy if Dst is larger than Src"
     );
```
This is no longer guaranteed to be compiled out. I do not believe we want this assert in this case.

tbu- replied:
Why is this not guaranteed to be compiled out?

nvzqz replied:
@tbu- because `size_of_val` for dynamically-sized types requires a runtime check since the type's size is not known at compile-time

nvzqz committed:
Change `std::mem::transmute_copy` to use `debug_assert!`

Code that uses this function is likely performance-sensitive, so we do
not want to emit an assert for dynamically-sized types in release mode.

the8472 commented:
Transmute only requires the bits to be valid for Dst, not necesarily that Dst is in a safe state.

Do we have any requirements on custom DSTs that their size must be computable at all times? One might get funny effects if such a type is trying to compute its size in the middle of unsafe code mucking with its bytes.

nvzqz commented:
@the8472 `size_of::<Dst>()` emits a compile-time constant, so I don't see how one would get funny effects at runtime

the8472 commented:
Hrrm, right, only the source is unsized. Still, the assert would try to compute the dynamic size of the source in debug builds of the standard library.

scottmcm commented:
Flipping this over to a libs-api reviewer since this needs an FCP due to being new stable capabilty
r? @m-ou-se

workingjubilee commented:
Tangentially: `ptr::read` will be const fn stable in 1.71.

scottmcm commented:
Out of curiosity, do you have any uses in mind for anything *but* `[u8]`?  For example, using this with trait objects feels much weirder than with byte slices.

But coming from `[u8]`, `bytes.as_ptr().cast::<Target>().read_unaligned()` would also work fine -- from a byte slice using the align-1 version is always fine -- and that makes me a bit unconvinced that changing `transmute_copy` is the right choice.

bors commented:
:umbrella: The latest upstream changes (presumably #115520) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Dylan-DPC commented:
@nvzqz would be nice to have a reply to https://github.com/rust-lang/rust/pull/112457#issuecomment-1603667722 and can resolve the conflicts if interested in moving this forward? thanks

Dylan-DPC commented:
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks
